agent/src/storage/mod.rs

```diff
@@ -1,80 +1,348 @@
 use std::path::PathBuf;
+use std::pin::Pin;
 
-use async_trait::async_trait;
+use aws_config::meta::region::RegionProviderChain;
+use aws_sdk_s3::config::Region;
+use axum::body::StreamBody;
+use bytes::Bytes;
 use once_cell::sync::OnceCell;
+use serval_client::ServalApiClient;
 use ssri::Integrity;
-use utils::errors::ServalError;
+use tokio::io::AsyncRead;
+use tokio_util::io::{ReaderStream, StreamReader};
+use utils::errors::{ServalError, ServalResult};
 use utils::mesh::ServalRole;
 use utils::structs::Manifest;
 
 pub mod blobs;
 pub use blobs::*;
 
-pub mod proxy;
-pub use proxy::*;
+pub mod bucket;
+pub use bucket::S3Storage;
 
 use crate::structures::MESH;
 
-/// If a node has local storage, this cell contains it.
-pub static STORAGE: OnceCell<BlobStore> = OnceCell::new();
+// A convenient alias for an often-used stream type.
+type SendableStream = Pin<Box<dyn AsyncRead + Send + 'static>>;
 
-/// Initialize our local storage.
-pub fn initialize(path: PathBuf) -> Result<(), ServalError> {
-    let store = BlobStore::new(path)?;
+/// Our fully-configured storage object, with all of its details hidden.
+pub static STORAGE: OnceCell<Storage> = OnceCell::new();
+
+/// Initialize our local storage and a proxy option if we have no storage ourselves.
+pub async fn initialize(path: Option<PathBuf>) -> ServalResult<()> {
+    let local = if let Some(blobpath) = path {
+        if let Ok(blob) = BlobStore::new(blobpath) {
```

Comment:
Noting a bit of discomfort here about tossing away the failure of `BlobStore::new` without doing anything about it. Should we propagate the error up to our caller? Log a warning? I am unsure.